File: step_15_graphs/15.5_mst_disjoint_set/P008_making_a_large_island/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE - Try flipping each 0, count resulting island
// Time: O(N^4)  |  Space: O(N^2)
// For each 0, temporarily set to 1 and BFS/DFS to count connected component
// ============================================================
class BruteForce {
    private static int N;

    public static int solve(int[][] grid) {
        N = grid.length;
        int maxSize = 0;
        // Check if already all 1s
        boolean hasZero = false;
        for (int[] row : grid) for (int v : row) if (v == 0) { hasZero = true; break; }
        if (!hasZero) return N * N;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == 0) {
                    grid[i][j] = 1;
                    maxSize = Math.max(maxSize, bfsSize(grid, i, j));
                    grid[i][j] = 0;
                }
            }
        }
        return maxSize;
    }

    private static int bfsSize(int[][] grid, int sr, int sc) {
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[N][N];
        q.offer(new int[]{sr, sc});
        visited[sr][sc] = true;
        int size = 0;
        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            size++;
            for (int[] d : dirs) {
                int nr = cur[0] + d[0], nc = cur[1] + d[1];
                if (nr >= 0 && nr < N && nc >= 0 && nc < N && grid[nr][nc] == 1 && !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    q.offer(new int[]{nr, nc});
                }
            }
        }
        return size;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL - DFS to label islands + check neighbors of each 0
// Time: O(N^2)  |  Space: O(N^2)
// Label each island with unique ID and precompute sizes.
// For each 0, sum up distinct neighboring island sizes + 1.
// ============================================================
class Optimal {
    private static int N;
    private static int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};

    public static int solve(int[][] grid) {
        N = grid.length;
        int[][] label = new int[N][N];
        Map<Integer, Integer> islandSize = new HashMap<>();
        int islandId = 2; // start from 2 (0=water,1=land unlabeled)

        // Label all islands
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == 1 && label[i][j] == 0) {
                    int size = dfs(grid, label, i, j, islandId);
                    islandSize.put(islandId, size);
                    islandId++;
                }
            }
        }

        int maxSize = islandSize.values().stream().mapToInt(x -> x).max().orElse(0);

        // Try flipping each 0
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == 0) {
                    Set<Integer> seen = new HashSet<>();
                    int total = 1; // the flipped cell itself
                    for (int[] d : dirs) {
                        int nr = i + d[0], nc = j + d[1];
                        if (nr >= 0 && nr < N && nc >= 0 && nc < N && label[nr][nc] > 0) {
                            int id = label[nr][nc];
                            if (seen.add(id)) total += islandSize.get(id);
                        }
                    }
                    maxSize = Math.max(maxSize, total);
                }
            }
        }
        return maxSize;
    }

    private static int dfs(int[][] grid, int[][] label, int r, int c, int id) {
        if (r < 0 || r >= N || c < 0 || c >= N || grid[r][c] != 1 || label[r][c] != 0) return 0;
        label[r][c] = id;
        int size = 1;
        for (int[] d : dirs) size += dfs(grid, label, r + d[0], c + d[1], id);
        return size;
    }
}

// ============================================================
// APPROACH 3: BEST - Union-Find with size tracking
// Time: O(N^2 * alpha(N^2))  |  Space: O(N^2)
// Use DSU to build islands, then for each 0 check union of neighbors
// ============================================================
class Best {
    static int[] parent, size;
    static int N;

    private static int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    private static void union(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return;
        if (size[px] < size[py]) { int t = px; px = py; py = t; }
        parent[py] = px;
        size[px] += size[py];
    }

    public static int solve(int[][] grid) {
        N = grid.length;
        parent = new int[N * N];
        size = new int[N * N];
        for (int i = 0; i < N * N; i++) { parent[i] = i; size[i] = 1; }

        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == 1) {
                    for (int[] d : dirs) {
                        int nr = i + d[0], nc = j + d[1];
                        if (nr >= 0 && nr < N && nc >= 0 && nc < N && grid[nr][nc] == 1) {
                            union(i * N + j, nr * N + nc);
                        }
                    }
                }
            }
        }

        int maxSize = 0;
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                if (grid[i][j] == 1) maxSize = Math.max(maxSize, size[find(i * N + j)]);

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == 0) {
                    Set<Integer> seen = new HashSet<>();
                    int total = 1;
                    for (int[] d : dirs) {
                        int nr = i + d[0], nc = j + d[1];
                        if (nr >= 0 && nr < N && nc >= 0 && nc < N && grid[nr][nc] == 1) {
                            int root = find(nr * N + nc);
                            if (seen.add(root)) total += size[root];
                        }
                    }
                    maxSize = Math.max(maxSize, total);
                }
            }
        }
        return maxSize;
    }
}

public class Solution {
    public static void main(String[] args) {
        int[][] g1 = {{1,0},{0,1}};
        int[][] g2 = {{1,1},{1,0}};
        int[][] g3 = {{1,1},{1,1}};

        System.out.println("Test 1 [[1,0],[0,1]] expected 3:");
        System.out.println("  Brute=" + BruteForce.solve(copyGrid(g1)) + " Optimal=" + Optimal.solve(copyGrid(g1)) + " Best=" + Best.solve(copyGrid(g1)));
        System.out.println("Test 2 [[1,1],[1,0]] expected 4:");
        System.out.println("  Brute=" + BruteForce.solve(copyGrid(g2)) + " Optimal=" + Optimal.solve(copyGrid(g2)) + " Best=" + Best.solve(copyGrid(g2)));
        System.out.println("Test 3 [[1,1],[1,1]] expected 4:");
        System.out.println("  Optimal=" + Optimal.solve(copyGrid(g3)) + " Best=" + Best.solve(copyGrid(g3)));
    }

    static int[][] copyGrid(int[][] g) {
        int[][] copy = new int[g.length][];
        for (int i = 0; i < g.length; i++) copy[i] = g[i].clone();
        return copy;
    }
}

File: step_15_graphs/15.5_mst_disjoint_set/P008_making_a_large_island/solution.py
"""Problem: Making a Large Island
Difficulty: HARD | XP: 50

Flip at most one 0 to 1, find the largest possible island.
Island = connected component of 1s (4-directional).
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE - Try flipping each 0, BFS to measure
# Time: O(N^4)  |  Space: O(N^2)
# ============================================================
def brute_force(grid: List[List[int]]) -> int:
    n = len(grid)
    dirs = ((0,1),(0,-1),(1,0),(-1,0))

    def bfs_size(g, sr, sc):
        from collections import deque
        q = deque([(sr, sc)])
        visited = {(sr, sc)}
        size = 0
        while q:
            r, c = q.popleft()
            size += 1
            for dr, dc in dirs:
                nr, nc = r+dr, c+dc
                if 0 <= nr < n and 0 <= nc < n and g[nr][nc] == 1 and (nr,nc) not in visited:
                    visited.add((nr,nc))
                    q.append((nr,nc))
        return size

    has_zero = any(grid[i][j] == 0 for i in range(n) for j in range(n))
    if not has_zero:
        return n * n

    max_size = 0
    for i in range(n):
        for j in range(n):
            if grid[i][j] == 0:
                grid[i][j] = 1
                max_size = max(max_size, bfs_size(grid, i, j))
                grid[i][j] = 0
    return max_size


# ============================================================
# APPROACH 2: OPTIMAL - DFS label islands + check each 0
# Time: O(N^2)  |  Space: O(N^2)
# Label each island with unique ID, precompute sizes.
# For each 0, sum sizes of distinct neighboring islands + 1.
# ============================================================
def optimal(grid: List[List[int]]) -> int:
    n = len(grid)
    label = [[0]*n for _ in range(n)]
    island_size = {}
    island_id = 2
    dirs = ((0,1),(0,-1),(1,0),(-1,0))

    def dfs(r, c, iid):
        if not (0 <= r < n and 0 <= c < n) or grid[r][c] != 1 or label[r][c] != 0:
            return 0
        label[r][c] = iid
        return 1 + sum(dfs(r+dr, c+dc, iid) for dr, dc in dirs)

    for i in range(n):
        for j in range(n):
            if grid[i][j] == 1 and label[i][j] == 0:
                sz = dfs(i, j, island_id)
                island_size[island_id] = sz
                island_id += 1

    max_size = max(island_size.values(), default=0)

    for i in range(n):
        for j in range(n):
            if grid[i][j] == 0:
                seen = set()
                total = 1
                for dr, dc in dirs:
                    nr, nc = i+dr, j+dc
                    if 0 <= nr < n and 0 <= nc < n and label[nr][nc] > 0:
                        lid = label[nr][nc]
                        if lid not in seen:
                            seen.add(lid)
                            total += island_size[lid]
                max_size = max(max_size, total)
    return max_size


# ============================================================
# APPROACH 3: BEST - Union-Find with component sizes
# Time: O(N^2 * alpha(N^2))  |  Space: O(N^2)
# ============================================================
def best(grid: List[List[int]]) -> int:
    n = len(grid)
    parent = list(range(n * n))
    size = [1] * (n * n)
    dirs = ((0,1),(0,-1),(1,0),(-1,0))

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(x, y):
        px, py = find(x), find(y)
        if px == py:
            return
        if size[px] < size[py]:
            px, py = py, px
        parent[py] = px
        size[px] += size[py]

    for i in range(n):
        for j in range(n):
            if grid[i][j] == 1:
                for dr, dc in dirs:
                    nr, nc = i+dr, j+dc
                    if 0 <= nr < n and 0 <= nc < n and grid[nr][nc] == 1:
                        union(i*n+j, nr*n+nc)

    max_size = max((size[find(i*n+j)] for i in range(n) for j in range(n) if grid[i][j] == 1), default=0)

    for i in range(n):
        for j in range(n):
            if grid[i][j] == 0:
                seen = set()
                total = 1
                for dr, dc in dirs:
                    nr, nc = i+dr, j+dc
                    if 0 <= nr < n and 0 <= nc < n and grid[nr][nc] == 1:
                        root = find(nr*n+nc)
                        if root not in seen:
                            seen.add(root)
                            total += size[root]
                max_size = max(max_size, total)
    return max_size


if __name__ == "__main__":
    import copy
    tests = [
        ([[1,0],[0,1]], 3),
        ([[1,1],[1,0]], 4),
        ([[1,1],[1,1]], 4),
        ([[0,0],[0,0]], 1),
    ]
    for grid, expected in tests:
        bf = brute_force(copy.deepcopy(grid))
        opt = optimal(copy.deepcopy(grid))
        be = best(copy.deepcopy(grid))
        status = "OK" if bf == opt == be == expected else "FAIL"
        print(f"[{status}] grid={grid} -> Brute={bf}, Optimal={opt}, Best={be} (expected={expected})")
